### PR TITLE
524 - Don't display usage() information for commands which run successfully

### PR DESCRIPTION
**Pull request by kennsippell**
# Description

#524 

Code:

- Moves all code that touches `actions` into function `buildActions`. Seems important to fully build that list before it is used - so that moves some validate-forms code earlier into main.
- ~`main.js` should be the only module to decide when usage information is displayed~

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: src/lib/main.js [hunk 1]
```diff
@@ -244,5 +199,58 @@ module.exports = async (argv, env) => {
   }
 };
 
+function buildActions(cmdArgs, skipValidate) {
```
Comment (jkuester) [inline src/lib/main.js:202]: Awww you ruined the great big run-on function that we had going with your smaller, well-named, easy-to-read function that encapsulates a distinct subset of the functionality....  :laughing:  :heart:

**Review by jkuester — COMMENTED** [review]
Added question in-line about possibly further improving the messaging for the end user (by eliminating an extra ERROR log).

File: src/lib/main.js [hunk 2]
```diff
@@ -128,66 +128,21 @@ module.exports = async (argv, env) => {
...
   // Initialize the environment
   //
   const projectName = fs.path.basename(pathToProject);
 
-  const apiUrl = getApiUrl(cmdArgs, env);
   const requiresInstance = actions.some(action => action.requiresInstance);
+  let apiUrl;
   if (requiresInstance) {
+    // getApiUrl makes log parsing statements, so call it only if needed
+    apiUrl = getApiUrl(cmdArgs, env);
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/main.js]: Ya that's better. Check it out.

File: src/lib/get-api-url.js [hunk 3]
```diff
@@ -26,8 +25,7 @@ const getApiUrl = (cmdArgs, env = {}) => {
     // See ./archiving-db.js
     instanceUrl = parseLocalUrl(env.COUCH_URL);
     if (instanceUrl.hostname !== 'localhost') {
-      error(`You asked to configure localhost, but the COUCH_URL env var is set to '${instanceUrl.hostname}'.  This may be a remote server.`);
-      return false;
+      throw Error(`--local was specified but COUCH_URL env var is set to '${instanceUrl.hostname}'.  Please use --url for remote servers.`);
```
Comment (jkuester) [inline src/lib/get-api-url.js:28]: :heart: Much improved message!

File: src/lib/main.js [hunk 4]
```diff
@@ -128,68 +128,19 @@ module.exports = async (argv, env) => {
...
   const requiresInstance = actions.some(action => action.requiresInstance);
-  if (requiresInstance) {
-    if (!apiUrl) {
-      throw new Error('Failed to obtain a url to the API');
-    }
+  const apiUrl = requiresInstance && getApiUrl(cmdArgs, env);
+  if (requiresInstance && !apiUrl) {
+    usage();
+    throw new Error('Failed to obtain a url to the API');
   }
```
Comment (jkuester) [inline src/lib/main.js:144]: Is it even possible to hit this case now? I think `getApiUrl` always return a value or throw an exception right?

**Review by jkuester — APPROVED** [review]
One minor question, but otherwise we are good to go!

File: src/lib/main.js (hunk 4 above)
Comment (kennsippell) [reply src/lib/main.js:144]: i don't believe it's possible as-written.

**Review by jkuester — APPROVED** [review]
@kennsippell can we merge these changes?

**Comment by garethbowen** [comment]
@kennsippell Looks like this passed AT. Merging is blocked because we've since added a Node 18 build as required. Would you mind rebasing to pick up the latest GH build file, then assuming everything passes, feel free to merge.

**Pull request merged**